src/Application/GameLuaBindings.h

```diff
@@ -1,16 +1,69 @@
 #pragma once
 
 #include <memory>
-#include <sol/forward.hpp>
+#include <unordered_map>
+#include <functional>
+#include <string>
+#include <sol/sol.hpp>
 
 struct lua_State;
 
+class Character;
+
 class GameLuaBindings {
  public:
     GameLuaBindings();
     ~GameLuaBindings();
     void init(lua_State *lua);
 
  private:
+    void _registerAudioBindings(sol::state_view &luaState, sol::table &table);
+    void _registerRenderBindings(sol::state_view &luaState, sol::table &table);
+    void _registerGameBindings(sol::state_view &luaState, sol::table &table);
+    void _registerPartyBindings(sol::state_view &luaState, sol::table &table);
+    void _registerItemBindings(sol::state_view &luaState, sol::table &table);
+    void _registerSerializationBindings(sol::state_view &luaState, sol::table &table);
+    void _registerEnums(sol::state_view &luaState, sol::table &table);
+
     std::unique_ptr<sol::state_view> _luaState;
+
+    // A helper class used to fill a lua table with all the requested information
+    template<typename ItemType>
+    class ItemQueryTable {
+     public:
+        explicit ItemQueryTable(sol::state_view &luaState) : _luaState(luaState) {}
+
+        template<typename T>
+        void add(const std::string_view &key, T&& query) {
+            _mapping.insert(std::make_pair(
+                key, [this, query](auto &item) { return sol::make_object(_luaState, query(item)); }
+            ));
+        }
+
+        /** Create a lua table by extracting the information from the provided item. */
+        sol::object createTable(const ItemType &item, const sol::object &queryTable) {
```

Comment:
A question on type safety. If I were writing this function, I'd be checking for the type of `queryTable` and only accept `nil` and `sol::table`, and throw otherwise. Am I missing smth, is this not how things are usually done in Lua?

Reply:
I underestimated the power of sol2. I can accept an optional as argument and I can even tell I expect a specific type ( in this case a vector of string )

```cpp
typedef std::optional<std::vector<std::string_view>> Table;
sol::object functionExposedInLua(Table queryTable);
```